java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+        intent.getParcelableExtra<ResultReceiver>(
+            OPERATION_RECEIVER
+        )?.send(0, bundle)
+    }
+
+    internal enum class Operation {
+        START_ARC,
+        STOP_ARC,
+        REGISTER_PARTICLE,
+        UNREGISTER_PARTICLE,
+        GET_REGISTERED_PARTICLES
+
+    }
+
+    companion object {
+        const val ARC_HOST_INTENT = "arcs.android.host.ARC_HOST"
+        private const val OPERATION = "OPERATION"
+        private const val OPERATION_ARG = "OPERATION_ARG"
+        private const val OPERATION_RECEIVER = "OPERATION_RECEIVER"
+        private const val OPERATION_RESULT = "OPERATION_RESULT"
+
+        internal fun createArcHostIntent(
+            operation: Operation,
+            component: ComponentName,
+            argument: Parcelable?
+        ): Intent = Intent(
+            ARC_HOST_INTENT
+        ).setComponent(component)
+            .putExtra(OPERATION, operation.ordinal)
+            .putExtra(
+                OPERATION_ARG, argument
+            )
+
+        @VisibleForTesting
+        fun onResult(intent: Intent, receiver: ResultReceiver?) =
+            intent.putExtra(
+                OPERATION_RECEIVER, receiver
+            )
+
+        @VisibleForTesting
+        fun getParticleIdentifierListResult(resultData: Bundle?): List<ParticleIdentifier>? =
+            resultData?.getParcelableArrayList<ParcelableParticleIdentifier>(
+                OPERATION_RESULT
+            )?.map { it -> it.actual }
+    }
+
+    /**
+     * An [ArcHost] stub that translates API calls to [Intent]s directed at a [Service] using
+     * [ArcHostHelper] to handle them.
+     *
+     * @property arcHostComponentName the [ComponentName] of the [Service]
+     * @property sender a callback used to fire the [Intent], overridable to allow testing.
+     */
+    class IntentArcHostAdapter(
+        private val arcHostComponentName: ComponentName,
+        private val sender: (Intent) -> Unit
+    ) : ArcHost {
+
+        override fun hostId() = arcHostComponentName.flattenToString()
+
+        override suspend fun registerParticle(particle: ParticleIdentifier) {
+            sendIntentToArcHostServiceForResult(
+                particle.createRegisterParticleIntent(arcHostComponentName)
+            )
+        }
+
+        override suspend fun unregisterParticle(particle: ParticleIdentifier) {
+            sendIntentToArcHostServiceForResult(
+                particle.createUnregisterParticleIntent(arcHostComponentName)
+            )
+        }
+
+        override suspend fun registeredParticles(): List<ParticleIdentifier> =
+            sendIntentToArcHostServiceForResult(
+                arcHostComponentName.createGetRegisteredParticlesIntent()
+            ) {
+                it?.let { list ->
+                    list as List<*>
+                    list.map {
+                        (it as ParcelableParticleIdentifier).actual
+                    }
+                }
+            } ?: emptyList()
```

Comment:
I think there's enough going on here, that for readability reasons: it makes sense to *not* have it be a function expression.